### PR TITLE
mgr/dashboard: validate username while creation

### PR DESCRIPTION
**Pull request by s0nea**
When creating a user the username is not checked if it's valid from the Ceph perspective (`CephString`). The commit adds a decorator to check if the input values in the API are valid from the Ceph perspective by calling the `valid()` method of the Ceph-defined datatypes (`ceph_argparse.py`).

Fixes: https://tracker.ceph.com/issues/46548
Signed-off-by: Tatjana Dehler <tdehler@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/__init__.py [hunk 1]
```diff
@@ -975,3 +976,19 @@ def allow_empty_body(func):  # noqa: N802
     except (AttributeError, KeyError):
         func._cp_config = {'tools.json_in.force': False}
     return func
+
+
+def validate_ceph_input(validations):
```
Comment (votdev) [inline src/pybind/mgr/dashboard/controllers/__init__.py:981]:
```suggestion
def validate_ceph_type(validations):
```

File: src/pybind/mgr/dashboard/controllers/__init__.py [hunk 2]
```diff
@@ -975,3 +976,19 @@ def allow_empty_body(func):  # noqa: N802
...
+    def decorator(func):
+        @wraps(func)
+        def validate_args(*args, **kwargs):
+            input_values = kwargs
+            for key, ceph_type in validations:
+                try:
+                    ceph_type.valid(input_values[key])
+                except ArgumentFormat as e:
+                    raise DashboardException(msg=e,
+                                             code='ceph_input_not_valid')
```
Comment (votdev) [inline src/pybind/mgr/dashboard/controllers/__init__.py:991]:
```suggestion
                                             code='ceph_type_not_valid')
```

**Comment by epuertat** [comment]
@s0nea regarding `dashboard/controllers/__init__.py:19: error: Cannot find implementation or library stub for module named 'ceph_argparse'`, you'll need to add `src/pybind` (or the relative path `../..`) to the `PYTHONPATH` in `mgr/tox.ini` and `dashboard/tox.ini`

**Comment by s0nea** [comment]
jenkins test api

**Comment by s0nea** [comment]
jenkins test api

**Comment by s0nea** [comment]
jenkins test api

**Comment by s0nea** [comment]
jenkins test api

**Comment by s0nea** [comment]
So... I guess that's the reason for the failing API test job:

```
2020-11-10T11:40:32.464+0000 7fc3d0d07200 -1 mgr[py] Module not found: 'dashboard'
2020-11-10T11:40:32.464+0000 7fc3d0d07200 -1 mgr[py] Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/__init__.py", line 55, in <module>
    from .module import Module, StandbyModule  # noqa: F401
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/module.py", line 21, in <module>
    from .controllers import generate_routes, json_error_page
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/__init__.py", line 20, in <module>
    from ceph_argparse import ArgumentFormat  # type: ignore
ModuleNotFoundError: No module named 'ceph_argparse'
```

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
```
jenkins-build@braggi17:~/build/workspace/ceph-api/build$ /home/jenkins-build/build/workspace/ceph-api/build/bin/ceph-mds -i c -c /home/jenkins-build/build/workspace/ceph-api/build/ceph.conf -f --debug-monc=10 --log-to-stderr=1
2020-11-13T15:22:43.832+0000 7fd3b9d8b600 -1 WARNING: all dangerous and experimental features are enabled.
2020-11-13T15:22:43.832+0000 7fd3b9d8b600 10 monclient: get_monmap_and_config
2020-11-13T15:22:43.832+0000 7fd3b9d8b600 10 monclient: build_initial_monmap
2020-11-13T15:22:43.832+0000 7fd3b9d8b600 10 monclient: monmap:
epoch 0
fsid 005e60df-59b2-42bb-ba63-ba187061afc4
last_changed 2020-11-13T15:22:43.834226+0000
created 2020-11-13T15:22:43.834226+0000
min_mon_release 0 (unknown)
election_strategy: 1
0: [v2:172.21.2.17:40895/0,v1:172.21.2.17:40896/0] mon.noname-a
1: [v2:172.21.2.17:40897/0,v1:172.21.2.17:40898/0] mon.noname-b
2: [v2:172.21.2.17:40899/0,v1:172.21.2.17:40900/0] mon.noname-c

2020-11-13T15:22:43.832+0000 7fd3b9d8b600 10 monclient: init
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient: _reopen_session rank -1
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient: _add_conns ranks=[0,1,2]
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): picked mon.noname-a con 0x557a5f364800 addr [v2:172.21.2.17:40895/0,v1:172.21.2.17:40896/0]
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): picked mon.noname-b con 0x557a5f364c00 addr [v2:172.21.2.17:40897/0,v1:172.21.2.17:40898/0]
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): picked mon.noname-c con 0x557a5f365000 addr [v2:172.21.2.17:40899/0,v1:172.21.2.17:40900/0]
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): start opening mon connection
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): start opening mon connection
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): start opening mon connection
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): _renew_subs
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient(hunting): authenticate will time out at 2020-11-13T15:27:43.838123+0000
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): get_auth_request con 0x557a5f364800 auth_method 0
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): get_auth_request method 2 preferred_modes [2,1]
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): _init_auth method 2
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): get_auth_request con 0x557a5f365000 auth_method 0
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): get_auth_request method 2 preferred_modes [2,1]
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): _init_auth method 2
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): get_auth_request con 0x557a5f364c00 auth_method 0
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): get_auth_request method 2 preferred_modes [2,1]
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): _init_auth method 2
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): handle_auth_reply_more payload 9
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): handle_auth_reply_more payload_len 9
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): handle_auth_reply_more responding with 36 bytes
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): handle_auth_reply_more payload 9
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): handle_auth_reply_more payload_len 9
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): handle_auth_reply_more responding with 36 bytes
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): handle_auth_reply_more payload 9
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): handle_auth_reply_more payload_len 9
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): handle_auth_reply_more responding with 36 bytes
2020-11-13T15:22:43.836+0000 7fd3a9a42700 10 monclient(hunting): handle_auth_bad_method old_auth_method 2 result (13) Permission denied allowed_methods [2]
2020-11-13T15:22:43.836+0000 7fd3a9a42700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
2020-11-13T15:22:43.836+0000 7fd3aa243700 10 monclient(hunting): handle_auth_bad_method old_auth_method 2 result (13) Permission denied allowed_methods [2]
2020-11-13T15:22:43.836+0000 7fd3aa243700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient(hunting): handle_auth_bad_method old_auth_method 2 result (13) Permission denied allowed_methods [2]
2020-11-13T15:22:43.836+0000 7fd3a9241700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient: _finish_hunting -13
2020-11-13T15:22:43.836+0000 7fd3a9241700  1 monclient: no mon sessions established
2020-11-13T15:22:43.836+0000 7fd3a9241700 10 monclient: _finish_auth -13
2020-11-13T15:22:43.836+0000 7fd3b9d8b600 10 monclient: shutdown
failed to fetch mon config (--no-mon-config to skip)
```

**Comment by tchaikov** [comment]
```
2020-11-13T14:15:36.611+0000 7f7900167200 -1 mgr[py] Module not found: 'dashboard'
2020-11-13T14:15:36.611+0000 7f7900167200 -1 mgr[py] Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/__init__.py", line 55, in <module>
    from .module import Module, StandbyModule  # noqa: F401
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/module.py", line 21, in <module>
    from .controllers import generate_routes, json_error_page
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/__init__.py", line 20, in <module>
    from ceph_argparse import ArgumentFormat  # type: ignore
ModuleNotFoundError: No module named 'ceph_argparse'

2020-11-13T14:15:36.611+0000 7f7900167200 -1 mgr[py] Class not found in module 'dashboard'
```

**Comment by tchaikov** [comment]
@s0nea hi Tatjana, please see #38063. i think we just need to add `src/pybind` to `PYTHONPATH`.

**Comment by s0nea** [comment]
> @s0nea hi Tatjana, please see #38063. i think we just need to add `src/pybind` to `PYTHONPATH`.

Thank you very much for your help @tchaikov!

**Review by tchaikov — APPROVED** [review]

**Review by alfonsomthd — APPROVED** [review]

**Comment by s0nea** [comment]
The QA test run was successful: https://pulpito.ceph.com/tdehler-2020-11-17_10:07:21-rados:dashboard-wip-tdehler-testing-37740-distro-basic-smithi/

**Pull request merged**